File: mvp-java/src/main/java/com/wf/mvp/java/view/IActivity.java
package com.wf.mvp.java.view;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.Message;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Toast;

import com.wf.mvp.java.customize.UiHandler;
import com.wf.mvp.java.impl.Ui;
import com.wf.mvp.java.presenter.DefaultIPresenter;
import com.wf.mvp.java.presenter.IPresenter;

import java.lang.reflect.ParameterizedType;

/**
 * MvpMode -> com.wf.mvp.java.view -> IActivity
 *
 * @Author: wf-pc
 * @Date: 2020-05-09 21:11
 * <p>
 * Mvp's v-layer basic Activity, used to bind presenter and implement common functions.
 */
public abstract class IActivity<P extends IPresenter> extends Activity implements Ui {

    /**
     * The instance object of P layer.
     */
    protected P mPresenter;

    /**
     * The instance object of UI thread handler.
     */
    protected UiHandler mUiHandler;

    private Toast mToast;
    private ProgressDialog mLoading;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        beforeCreate();
        super.onCreate(savedInstanceState);
        setContentView(bindLayoutId());
        initView(getIntent());
        attachPresenter();
        initListener();
        initData();
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        setIntent(intent);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        detachPresenter();
        release();
    }


    /**
     * Binding P layer instance object.
     * Obtain the generic class type through reflection, and then instantiate the Presenter object.
     */
    @SuppressWarnings("unchecked")
    @Override
    public void attachPresenter() {
        mUiHandler = new UiHandler();
        mUiHandler.attachRefs(this);
        try {
            ParameterizedType type = (ParameterizedType) this.getClass().getGenericSuperclass();
            Class<P> tClass = (Class<P>) type.getActualTypeArguments()[0];
            mPresenter = tClass.newInstance();
        } catch (Exception e) {
            mPresenter = (P) new DefaultIPresenter();
        }
        mPresenter.attachView(this, mUiHandler);
    }

    /**
     * Unbind the P layer instance object.
     */
    @Override
    public void detachPresenter() {
        if (mUiHandler != null) {
            mUiHandler.detachRefs();
            mUiHandler = null;
        }
        if (mPresenter != null) {
            mPresenter.detachView();
            mPresenter = null;
        }
    }


    /**
     * Used to do something before Create.
     */
    protected void beforeCreate() {}

    /**
     * Used to bind layout files.
     */
    protected abstract int bindLayoutId();

    /**
     * Used to initialize view.
     */
    protected abstract void initView(Intent intent);

    /**
     * Used to initialize listener.
     */
    protected abstract void initListener();

    /**
     * Used to initialize data.
     */
    protected abstract void initData();

    /**
     * Used to release resources.
     */
    protected abstract void release();

    /**
     * It is not defined as an abstract function here,
     * it is recommended to put the message to the Presenter layer for processing.
     */
    @Override
    public void onHandleMessage(Message msg) {}


    @Override
    public float getDimension(int resId) {
        return getResources().getDimension(resId);
    }

    @SuppressLint("ShowToast")
    @Override
    public void showToast(String message) {
        if (mToast == null) {
            mToast = Toast.makeText(this, "", Toast.LENGTH_SHORT);
        }
        mToast.setText(message);
        mToast.show();
    }

    @SuppressLint("ShowToast")
    @Override
    public void showLongToast(String message) {
        if (mToast == null) {
            mToast = Toast.makeText(this, "", Toast.LENGTH_LONG);
        }
        mToast.setText(message);
        mToast.show();
    }

    @Override
    public void showLoading() {
        showLoading("");
    }

    @Override
    public void showLoading(String message) {
        mLoading = ProgressDialog.show(this, "", message);
    }

    @Override
    public void hideLoading() {
        if (mLoading != null && mLoading.isShowing()) {
            mLoading.dismiss();
        }
    }

    @Override
    public void toggleKeyboard() {
        InputMethodManager imm = (InputMethodManager) this.getSystemService(Context.INPUT_METHOD_SERVICE);
        if (imm != null && this.getCurrentFocus() != null
                && this.getCurrentFocus().getWindowToken() != null) {
            imm.toggleSoftInputFromWindow(this.getCurrentFocus().getWindowToken(),
                    0, InputMethodManager.HIDE_NOT_ALWAYS);
        }
    }

    @Override
    public void showKeyboard(View view) {
        if (view == null) return;
        view.requestFocus();
        InputMethodManager imm = (InputMethodManager) this.getSystemService(Context.INPUT_METHOD_SERVICE);
        if (imm != null) {
            imm.showSoftInput(view, InputMethodManager.SHOW_FORCED);
        }
    }

    @Override
    public void hideKeyboard() {
        InputMethodManager imm = (InputMethodManager) this.getSystemService(Context.INPUT_METHOD_SERVICE);
        if (imm != null && imm.isActive() && this.getCurrentFocus() != null
                && this.getCurrentFocus().getWindowToken() != null) {
            imm.hideSoftInputFromWindow(this.getCurrentFocus().getWindowToken(),
                    InputMethodManager.HIDE_NOT_ALWAYS);
        }
    }
}

File: mvp-kotlin/src/main/java/com/wf/mvp/kotlin/view/IActivity.kt
package com.wf.mvp.kotlin.view

import android.annotation.SuppressLint
import android.app.Activity
import android.app.ProgressDialog
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Message
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Toast
import com.wf.mvp.kotlin.customize.UiHandler
import com.wf.mvp.kotlin.impl.Presenter
import com.wf.mvp.kotlin.impl.Ui
import com.wf.mvp.kotlin.presenter.DefaultIPresenter
import java.lang.reflect.ParameterizedType

/**
 * MvpMode -> com.wf.mvp.kotlin.view -> IActivity
 * @Author: wf-pc
 * @Date: 2020-05-10 16:42
 * <p>
 * Mvp's v-layer basic Activity, used to bind presenter and implement common functions.
 */
abstract class IActivity<P : Presenter> : Activity(), Ui {

    /**
     * The instance object of P layer.
     */
    protected var mPresenter: P? = null

    /**
     * The instance object of UI thread handler.
     */
    protected var mUiHandler: UiHandler? = null

    private var mToast: Toast? = null
    private var mLoading: ProgressDialog? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        beforeCreate()
        super.onCreate(savedInstanceState)
        setContentView(bindLayoutId())
        initView(intent)
        attachPresenter()
        initListener()
        initData()
    }

    override fun onNewIntent(intent: Intent?) {
        super.onNewIntent(intent)
        setIntent(intent)
    }

    override fun onDestroy() {
        super.onDestroy()
        detachPresenter()
        release()
    }


    /**
     * Binding P layer instance object.
     * Obtain the generic class type through reflection, and then instantiate the Presenter object.
     */
    @Suppress("UNCHECKED_CAST")
    override fun attachPresenter() {
        mUiHandler = UiHandler()
        mUiHandler?.attachRefs(this)
        mPresenter = try {
            val type = this.javaClass.genericSuperclass as ParameterizedType
            val tClass: Class<P> = type.actualTypeArguments[0] as Class<P>
            tClass.newInstance()
        } catch (e: Exception) {
            DefaultIPresenter() as P
        }
        mPresenter?.attachView(this, mUiHandler)
    }

    /**
     * Unbind the P layer instance object.
     */
    override fun detachPresenter() {
        mUiHandler?.detachRefs()
        mPresenter?.detachView()
        mUiHandler = null
        mPresenter = null
    }


    /**
     * Used to do something before Create.
     */
    protected open fun beforeCreate() {}

    /**
     * Used to bind layout files.
     */
    protected abstract fun bindLayoutId(): Int

    /**
     * Used to initialize view.
     */
    protected abstract fun initView(intent: Intent?)

    /**
     * Used to initialize listener.
     */
    protected abstract fun initListener()

    /**
     * Used to initialize data.
     */
    protected abstract fun initData()

    /**
     * Used to release resources.
     */
    protected abstract fun release()

    /**
     * It is not defined as an abstract function here,
     * it is recommended to put the message to the Presenter layer for processing.
     */
    override fun onHandleMessage(msg: Message) {}


    override fun runOnUiThread(block: () -> Unit) {
        mUiHandler?.post(block)
    }

    override fun getDimension(resId: Int): Float {
        return resources?.getDimension(resId) ?: 0f
    }

    @SuppressLint("ShowToast")
    override fun showToast(message: String) {
        if (mToast == null) {
            mToast = Toast.makeText(this, "", Toast.LENGTH_SHORT)
        }
        mToast?.setText(message)
        mToast?.show()
    }

    @SuppressLint("ShowToast")
    override fun showLongToast(message: String) {
        if (mToast == null) {
            mToast = Toast.makeText(this, "", Toast.LENGTH_LONG)
        }
        mToast?.setText(message)
        mToast?.show()
    }

    override fun showLoading() {
        showLoading("")
    }

    override fun showLoading(message: String) {
        mLoading = ProgressDialog.show(this, "", message)
    }

    override fun hideLoading() {
        mLoading?.dismiss()
    }

    override fun toggleKeyboard() {
        val imm = this.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager?
        imm?.toggleSoftInputFromWindow(
            this.currentFocus?.windowToken,
            0,
            InputMethodManager.HIDE_NOT_ALWAYS
        )
    }

    override fun showKeyboard(view: View) {
        view.requestFocus()
        val imm = this.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager?
        imm?.showSoftInput(view, InputMethodManager.SHOW_FORCED)
    }

    override fun hideKeyboard() {
        val imm = this.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager?
        imm?.hideSoftInputFromWindow(
            this.currentFocus?.windowToken,
            InputMethodManager.HIDE_NOT_ALWAYS
        )
    }
}